In [58]:
# Load libraries and global parameters
packages =  c("lubridate","Hmisc", "ggplot2","directlabels", "dplyr", "tidyr", "data.table", 'corrplot', 'gridExtra', 'forecast', 'tseries', 'TSA', 'tibble', 'TTR', 'xts', 'dygraphs', 'assertthat')

my.install <- function(pkg, ...){
  if (!(pkg %in% installed.packages()[,1])) {
    install.packages(pkg)
  }
  return (library(pkg, ...))
}

purrr::walk(packages, my.install, character.only = TRUE, warn.conflicts = FALSE)

In [57]:
crime.data = read.csv('crime_report.csv')
str(crime.data)

'data.frame':	2829 obs. of  15 variables:
 $ report_year        : int  1975 1975 1975 1975 1975 1975 1975 1975 1975 1975 ...
 $ agency_code        : Factor w/ 69 levels "","AZ00717","AZ00723",..: 42 59 22 14 61 30 33 29 44 39 ...
 $ agency_jurisdiction: Factor w/ 69 levels "Albuquerque, NM",..: 1 2 3 4 5 6 7 8 9 10 ...
 $ population         : int  286238 112478 490584 116656 300400 642154 864100 616120 422276 262103 ...
 $ violent_crimes     : int  2383 278 8033 611 1215 1259 16086 11386 3350 1937 ...
 $ homicides          : int  30 5 185 7 33 25 259 119 63 68 ...
 $ rapes              : int  181 28 443 44 190 137 463 453 192 71 ...
 $ assaults           : int  1353 132 3518 389 463 347 6309 3036 755 976 ...
 $ robberies          : int  819 113 3887 171 529 750 9055 7778 2340 822 ...
 $ months_reported    : int  12 12 12 12 12 12 12 12 12 12 ...
 $ crimes_percapita   : num  833 247 1637 524 404 ...
 $ homicides_percapita: num  10.48 4.45 37.71 6 10.99 ...
 $ rapes_percapita    : num  6

In [3]:
#Overview of data
summary(crime.data)
str(crime.data)

  report_year    agency_code             agency_jurisdiction   population     
 Min.   :1975          :  41   Albuquerque, NM     :  41     Min.   : 100763  
 1st Qu.:1985   AZ00717:  41   Arlington, TX       :  41     1st Qu.: 377931  
 Median :1995   AZ00723:  41   Atlanta, GA         :  41     Median : 536614  
 Mean   :1995   AZ01003:  41   Aurora, CO          :  41     Mean   : 795698  
 3rd Qu.:2005   CA00109:  41   Austin, TX          :  41     3rd Qu.: 816856  
 Max.   :2015   CA01005:  41   Baltimore County, MD:  41     Max.   :8550861  
                (Other):2583   (Other)             :2583     NA's   :69       
 violent_crimes      homicides           rapes           assaults    
 Min.   :    154   Min.   :    1.0   Min.   :  15.0   Min.   :   15  
 1st Qu.:   3015   1st Qu.:   32.0   1st Qu.: 176.2   1st Qu.: 1467  
 Median :   5136   Median :   64.0   Median : 291.0   Median : 2597  
 Mean   :  29633   Mean   :  398.4   Mean   : 416.3   Mean   : 4405  
 3rd Qu.:   9058  

'data.frame':	2829 obs. of  15 variables:
 $ report_year        : int  1975 1975 1975 1975 1975 1975 1975 1975 1975 1975 ...
 $ agency_code        : Factor w/ 69 levels "","AZ00717","AZ00723",..: 42 59 22 14 61 30 33 29 44 39 ...
 $ agency_jurisdiction: Factor w/ 69 levels "Albuquerque, NM",..: 1 2 3 4 5 6 7 8 9 10 ...
 $ population         : int  286238 112478 490584 116656 300400 642154 864100 616120 422276 262103 ...
 $ violent_crimes     : int  2383 278 8033 611 1215 1259 16086 11386 3350 1937 ...
 $ homicides          : int  30 5 185 7 33 25 259 119 63 68 ...
 $ rapes              : int  181 28 443 44 190 137 463 453 192 71 ...
 $ assaults           : int  1353 132 3518 389 463 347 6309 3036 755 976 ...
 $ robberies          : int  819 113 3887 171 529 750 9055 7778 2340 822 ...
 $ months_reported    : int  12 12 12 12 12 12 12 12 12 12 ...
 $ crimes_percapita   : num  833 247 1637 524 404 ...
 $ homicides_percapita: num  10.48 4.45 37.71 6 10.99 ...
 $ rapes_percapita    : num  6

In [4]:
#Data Cleaning
crime.data[!complete.cases(crime.data),]
nrow(crime.data[!complete.cases(crime.data),])

report_year agency_code agency_jurisdiction        population
31   1975        KY05680     Louisville, KY                 NA    
69   1975                    United States                  NA    
100  1976        KY05680     Louisville, KY                 NA    
138  1976                    United States                  NA    
169  1977        KY05680     Louisville, KY                 NA    
207  1977                    United States                  NA    
238  1978        KY05680     Louisville, KY                 NA    
276  1978                    United States                  NA    
307  1979        KY05680     Louisville, KY                 NA    
345  1979                    United States                  NA    
376  1980        KY05680     Louisville, KY                 NA    
414  1980                    United States                  NA    
445  1981        KY05680     Louisville, KY                 NA    
483  1981                    United States                  NA    
514  1982        KY05680     Louisville, KY                 NA    
552  1982                    United States                  NA    
583  1983        KY05680     Louisville, KY                 NA    
621  1983                    United States                  NA    
652  1984        KY05680     Louisville, KY                 NA    
690  1984                    United States                  NA    
721  1985        KY05680     Louisville, KY                 NA    
759  1985                    United States                  NA    
790  1986        KY05680     Louisville, KY                 NA    
828  1986                    United States                  NA    
859  1987        KY05680     Louisville, KY                 NA    
897  1987                    United States                  NA    
922  1988        FL01602     Jacksonville, FL           645024    
928  1988        KY05680     Louisville, KY                 NA    
960  1988        FL02902     Tampa, FL                  293732    
966  1988                    United States                  NA    
...  ...         ...         ...                        ...       
2800 2015        NY02900     Nassau County, NY          1061175   
2801 2015        LANPD00     New Orleans, LA             393447   
2802 2015        NY03030     New York City, NY          8550861   
2803 2015        NJNPD00     Newark, NJ                  279110   
2804 2015        CA00109     Oakland, CA                 419481   
2805 2015        OK05506     Oklahoma City, OK           630621   
2806 2015        NB02802     Omaha, NE                   452252   
2807 2015        FL04804     Orlando, FL                 268438   
2808 2015        PAPEP00     Philadelphia, PA           1567810   
2809 2015        AZ00723     Phoenix, AZ                1559744   
2810 2015        PAPPD00     Pittsburgh, PA              306870   
2811 2015        OR02602     Portland, OR                615672   
2812 2015        MD01721     Prince George's County, MD  685191   
2813 2015        NC09201     Raleigh, NC                 438363   
2814 2015        CA03404     Sacramento, CA              489717   
2815 2015        UT01803     Salt Lake City, UT          191992   
2816 2015        TXSPD00     San Antonio, TX            1463586   
2817 2015        CA03711     San Diego, CA              1400467   
2818 2015        CA03801     San Francisco, CA           863782   
2819 2015        CA04313     San Jose, CA               1031458   
2820 2015        WASPD00     Seattle, WA                 683700   
2821 2015        MOSPD00     St. Louis, MO               317095   
2822 2015        NY05101     Suffolk County, NY         1341453   
2823 2015        FL02902     Tampa, FL                   364383   
2824 2015        AZ01003     Tucson, AZ                  529675   
2825 2015        OK07205     Tulsa, OK                   401520   
2826 2015        VA12800     Virginia Beach, VA          452797   
2827 2015        DCMPD00     Washington, DC         

[1] 141

The total records having missing values is 141. Upon suspection, the column months_reported has the most missing values. Since it is not significant in the research, it is removed. After removal, incomplete cases reduced to 76. These 76 records were then removed rather than imputed since the data used is very sensitive and is unethical to assume a crime rate for a particular area.

In [5]:
# remove the column month_reported
crime.data = select(crime.data, -c(months_reported))
nrow(crime.data[!complete.cases(crime.data),])

[1] 76

In [6]:
# remove records with missing values
crime.data = na.omit(crime.data)
summary(crime.data)

  report_year    agency_code        agency_jurisdiction   population     
 Min.   :1975   AZ00717:  41   Albuquerque, NM:  41     Min.   : 100763  
 1st Qu.:1985   AZ00723:  41   Arlington, TX  :  41     1st Qu.: 378014  
 Median :1995   AZ01003:  41   Atlanta, GA    :  41     Median : 537081  
 Mean   :1995   CA00109:  41   Aurora, CO     :  41     Mean   : 796458  
 3rd Qu.:2005   CA01005:  41   Austin, TX     :  41     3rd Qu.: 817083  
 Max.   :2015   CA01900:  41   Baltimore, MD  :  41     Max.   :8550861  
                (Other):2507   (Other)        :2507                      
 violent_crimes     homicides          rapes           assaults    
 Min.   :   154   Min.   :   1.0   Min.   :  15.0   Min.   :   15  
 1st Qu.:  2965   1st Qu.:  32.0   1st Qu.: 176.0   1st Qu.: 1467  
 Median :  5066   Median :  63.0   Median : 291.0   Median : 2597  
 Mean   :  8946   Mean   : 124.9   Mean   : 416.2   Mean   : 4405  
 3rd Qu.:  8804   3rd Qu.: 127.0   3rd Qu.: 465.0   3rd Qu.: 4556  


In [7]:
#Extracting crime columns from the data
crime.data.crime = crime.data[,c(5,6,7,8,9)]
summary(crime.data.crime)

 violent_crimes     homicides          rapes           assaults    
 Min.   :   154   Min.   :   1.0   Min.   :  15.0   Min.   :   15  
 1st Qu.:  2965   1st Qu.:  32.0   1st Qu.: 176.0   1st Qu.: 1467  
 Median :  5066   Median :  63.0   Median : 291.0   Median : 2597  
 Mean   :  8946   Mean   : 124.9   Mean   : 416.2   Mean   : 4405  
 3rd Qu.:  8804   3rd Qu.: 127.0   3rd Qu.: 465.0   3rd Qu.: 4556  
 Max.   :174542   Max.   :2245.0   Max.   :3899.0   Max.   :71030  
   robberies     
 Min.   :    83  
 1st Qu.:  1032  
 Median :  1939  
 Mean   :  4000  
 3rd Qu.:  3609  
 Max.   :107475  

In [52]:
#crime.data.crime = aggregate(crime.data[,c(5,6,7,8,9)], by=list(crime.data$agency_jurisdiction), FUN=sum)
#total = rowSums(crime.data.crime[,c(2,3,4,5,6)])
#crime.data.crime = cbind(crime.data.crime, total)
#crime.data.crime
total = rowSums(crime.data[,c(5,6,7,8,9)])
crime.data = cbind(crime.data, total)

In [ ]:
library(directlabels)
ggplot( data = crime.data, aes(x=report_year, y=total, group=agency_jurisdiction, color=agency_jurisdiction)) +
    geom_line() + 
    geom_dl(aes(label=agency_jurisdiction), method="last.points") 


#graph shows in R

In [59]:
#Add date column for ggplot time series
library(lubridate)
report_date <- as.Date(as.character(crime.data$report_year), format = "%Y")
cbind(crime.data, report_date)

report_year agency_code agency_jurisdiction        population
1    1975        NM00101     Albuquerque, NM             286238   
2    1975        TX22001     Arlington, TX               112478   
3    1975        GAAPD00     Atlanta, GA                 490584   
4    1975        CO00101     Aurora, CO                  116656   
5    1975        TX22701     Austin, TX                  300400   
6    1975        MD00301     Baltimore County, MD        642154   
7    1975        MDBPD00     Baltimore, MD               864100   
8    1975        MA01301     Boston, MA                  616120   
9    1975        NY01401     Buffalo, NY                 422276   
10   1975        NC06001     Charlotte, NC               262103   
11   1975        ILCPD00     Chicago, IL                3150000   
12   1975        OHCIP00     Cincinnati, OH              433367   
13   1975        OHCLP00     Cleveland, OH               659931   
14   1975        OHCOP00     Columbus, OH                572797   
15   1975        TXDPD00     Dallas, TX                  864665   
16   1975        CODPD00     Denver, CO                  508140   
17   1975        MI82349     Detroit, MI                1432444   
18   1975        TX07102     El Paso, TX                 369000   
19   1975        VA02901     Fairfax County, VA          487905   
20   1975        TX22012     Fort Worth, TX              381275   
21   1975        CA01005     Fresno, CA                  176500   
22   1975        HI00200     Honolulu, HI                705262   
23   1975        TXHPD00     Houston, TX                1372342   
24   1975        INIPD00     Indianapolis, IN            503411   
25   1975        FL01602     Jacksonville, FL            542792   
26   1975        MOKPD00     Kansas City, MO             489094   
27   1975        NV00201     Las Vegas, NV               249186   
28   1975        CA01941     Long Beach, CA              340758   
29   1975        CA01900     Los Angeles County, CA     1022888   
30   1975        CA01942     Los Angeles, CA            2729878   
...  ...         ...         ...                        ...       
2800 2015        NY02900     Nassau County, NY          1061175   
2801 2015        LANPD00     New Orleans, LA             393447   
2802 2015        NY03030     New York City, NY          8550861   
2803 2015        NJNPD00     Newark, NJ                  279110   
2804 2015        CA00109     Oakland, CA                 419481   
2805 2015        OK05506     Oklahoma City, OK           630621   
2806 2015        NB02802     Omaha, NE                   452252   
2807 2015        FL04804     Orlando, FL                 268438   
2808 2015        PAPEP00     Philadelphia, PA           1567810   
2809 2015        AZ00723     Phoenix, AZ                1559744   
2810 2015        PAPPD00     Pittsburgh, PA              306870   
2811 2015        OR02602     Portland, OR                615672   
2812 2015        MD01721     Prince George's County, MD  685191   
2813 2015        NC09201     Raleigh, NC                 438363   
2814 2015        CA03404     Sacramento, CA              489717   
2815 2015        UT01803     Salt Lake City, UT          191992   
2816 2015        TXSPD00     San Antonio, TX            1463586   
2817 2015        CA03711     San Diego, CA              1400467   
2818 2015        CA03801     San Francisco, CA           863782   
2819 2015        CA04313     San Jose, CA               1031458   
2820 2015        WASPD00     Seattle, WA                 683700   
2821 2015        MOSPD00     St. Louis, MO               317095   
2822 2015        NY05101     Suffolk County, NY         1341453   
2823 2015        FL02902     Tampa, FL                   364383   
2824 2015        AZ01003     Tucson, AZ                  529675   
2825 2015        OK07205     Tulsa, OK                   401520   
2826 2015        VA12800     Virginia Beach, VA          452797   
2827 2015        DCMPD00     Washington, DC         